insights/specs/datasources/system_user_dirs.py

```diff
@@ -72,33 +74,48 @@ def system_user_dirs(broker):
     r"""
     Custom datasource for CVE-2021-35937, CVE-2021-35938, and CVE-2021-35939.
 
-    It collects package names from the ``rpm -qa --qf="[%{=NAME}; %{FILEMODES:perms};
-    %{FILEUSERNAME}; %{FILEGROUPNAME}\n]" command``, if the package has
-    at least one directory which is writable by a specific user/group or the others.
+    It collects package names from the ``rpm -qa --nosignature --qf="[%{=NAME}; %{FILENAMES};
+    %{FILEMODES:perms}; %{FILEUSERNAME}; %{FILEGROUPNAME}\n]" command``.
+
+    The output is a sorted list of all packages, which have at least one directory with files
+    inside, and this directory is writable by a specific user/group or the others.
 
     Raises:
         SkipComponent: Raised if no data is available
 
     Returns:
         List[str]: Sorted list of package names
     """
-    users = get_users()
-    groups = get_groups(users)
-    packages = set()
     content = broker[LocalSpecs.rpm_args].content
 
     if not content or "command not found" in content[0]:
         raise SkipComponent
 
+    users = get_users()
+    groups = get_groups(users)
+
+    dirs = {}
+    packages = set()
+
+    # Stores a writeable directory with its package
     for line in content:
-        name, perms, user, group = line.split("; ")
+        pkg_name, path_name, perms, user, group = line.split("; ")
         if perms[0] == "d":
             user_w = user in users and perms[2] == "w"
             group_w = group in groups and perms[5] == "w"
             others_w = perms[8] == "w"
 
             if user_w or group_w or others_w:
-                packages.add(name)
+                dirs[path_name] = pkg_name
+
+    # Stores a package if its associated file is in a writable directory
+    for line in content:
+        _, path_name, perms, _, _ = line.split("; ")
+        if perms[0] != "d":
+            dir_name = path_name.rsplit('/', 1)[0]
+
+            if dir_name in dirs:
+                packages.add(dirs[dir_name])
```

Comment:
From Ryan's script's result, the `loop` is the most time-consuming part on that host, I think it's not so good to iterate the content twice, how about changing to an equivalent set-comprehension to a built-in query against the dict?
```python
    dir_package = {}
    dirs = set()
    packages  = set()
    # Stores a writeable directory with its package
    for line in content:
        pkg_name, path_name, perms, user, group = line.split("; ")
        if perms[0] == "d":
            user_w = user in users and perms[2] == "w"
            group_w = group in groups and perms[5] == "w"
            others_w = perms[8] == "w"
            dirs.add(path_name.rsplit('/', 1)[0])
            if user_w or group_w or others_w:
                dir_package[path_name] = pkg_name

    # Stores a package if its associated file is in a writable directory
    packages.update(dir_package[dir] for dir in dirs if dir in dir_package)
```

Reply:
This code is not correct although the idea has some merit. The second loop looks at all files not only dirs. So dirs.add line needs to be outside the if perms[0] == "d": block.

Reply:
Yeap, If you are right, then [the original code](https://github.com/RedHatInsights/insights-core/pull/3586/files#diff-62ba94046189d470d928eedfbab6784c6b0cc238e80c112289fb484791bbbe09R111-R115) must be wrong. My update strictly followed the original code and they must be equivalent.  @jobselko, If you think it's okay to update per my comment, please pay attention to the comment from @ffesti

Reply:
Oops, my bad, I didn't note that the two `if` statement is opposite.

The correct code should be:

```
    dir_package = {}
    dirs = set()
    packages  = set()
    # Stores a writeable directory with its package
    for line in content:
        pkg_name, path_name, perms, user, group = line.split("; ")
        if perms[0] == "d":
            user_w = user in users and perms[2] == "w"
            group_w = group in groups and perms[5] == "w"
            others_w = perms[8] == "w"
            if user_w or group_w or others_w:
                dir_package[path_name] = pkg_name
        else:
            dirs.add(path_name.rsplit('/', 1)[0])

    # Stores a package if its associated file is in a writable directory
    packages.update(dir_package[dir] for dir in dirs if dir in dir_package)
```

Sorry about that.